server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -449,4 +461,51 @@ private Decision decideCanForceAllocateForVacate(ShardRouting shardRouting, Rout
         assert target != null : "Target node is not found";
         return allocation.deciders().canForceAllocateDuringReplace(shardRouting, target, allocation);
     }
+
+    private OrderedNodesShardsIterator createOrderedNodesShardsIterator() {
+        return new OrderedNodesShardsIterator();
+    }
+
+    private class OrderedNodesShardsIterator implements Iterator<ShardRouting> {
```

Comment:
I am planning to move this to a top-level class and add some unit tests for this